File: 053-正则表达式匹配/053-match.cpp
class Solution {
public:
    bool match(char* s, char* p)
    {
        if (*p == '\0')         //  正则p到底末尾时
        {
            return !(*s);       //  如果串s页到达末尾，则匹配成功
        }

        int slen = strlen(s), plen = strlen(p);

        if (plen == 1           //  如果正则串只有一个长度
        || *(p + 1) != '*')     //  如果匹配*
        {
            //
            return slen && (p[0] == '.' || *s == *p)
                && match(s + 1, p + 1);
        }
        else
        {
            //  匹配一个字符
            while (*s != '\0' && (*p == '.' || *s == *p))
            {
                if (match(s++, p + 2))
                {
                    return true;
                }
            }
        }

        //  匹配0个字符
        return match(s, p + 2);
    }
};

File: 053-正则表达式匹配/053-match.py
# -*- coding:utf-8 -*-
class Solution:
    # s, pattern都是字符串
    def match(self, s, pattern):
        # write code here
        dp=[[False for _ in range(len(pattern)+1)] for _ in range(len(s)+1)]
        dp[0][0]=True
        for i in range(1,len(s)+1):
            dp[i][0]=False
        for i in range(2,len(pattern)+1):
            dp[0][i]= (pattern[i-1]=='*') and dp[0][i-2]
        for i in range(1,len(s)+1):
            for j in range(1,len(pattern)+1):
                if pattern[j-1]!='*':
                    dp[i][j]=dp[i-1][j-1] and (s[i-1]==pattern[j-1] or '.'==pattern[j-1])
                else:
                    dp[i][j]= dp[i][j-2] or (s[i-1]==pattern[j-2] or '.'==pattern[j-2]) and dp[i-1][j]
        return dp[-1][-1]
